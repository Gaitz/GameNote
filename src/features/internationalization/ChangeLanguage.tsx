import { ChangeEventHandler } from "react"
import { useRouter } from "next/router"
import { useTranslation } from "next-i18next"
import { Select } from "@chakra-ui/react"
import nookies from "nookies"

type supportedLocales = "en" | "zh-TW"
type supportedLocalesNameMapping = { [locale in supportedLocales]: string }

const supportedLocalesName: supportedLocalesNameMapping = {
  en: "English",
  "zh-TW": "繁體中文"
}

// [reference](https://nextjs.org/docs/advanced-features/i18n-routing#leveraging-the-next_locale-cookie)
const NEXT_LOCALE_COOKIE = "NEXT_LOCALE"

export function ChangeLanguage() {
  const { t } = useTranslation("login")
  const router = useRouter()

  const languageChangeHandler: ChangeEventHandler<HTMLSelectElement> = (
    event
  ) => {
    const selectedLanguage = event.target.value
    const isChanged = selectedLanguage !== router.locale

    if (selectedLanguage && isChanged) {
      const currentPath = router.pathname
      router.push(currentPath, currentPath, { locale: selectedLanguage })
      nookies.set(null, NEXT_LOCALE_COOKIE, selectedLanguage)
    }
  }

  return (
    <Select
      placeholder={t("languageSwitch")}
      onChange={languageChangeHandler}
      width="max-content"
    >
      {Object.entries(supportedLocalesName).map(([locale, name]) => {
        return (
          <option key={locale} value={locale}>
            {name}
          </option>
        )
      })}
    </Select>
  )
}
