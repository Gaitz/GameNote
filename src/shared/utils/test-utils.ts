import { render } from "@testing-library/react"
import store from "game-note/shared/store"
import theme from "game-note/shared/styles/theme"

// const AllTheProviders = ({ children }) => {
//   return (
//     <ReduxProvider store={store}>
//       <ChakraProvider theme={theme}>{children}</ChakraProvider>
//     </ReduxProvider>
//   )
// }

// const customRender = (ui: ReactElement, options?: object) =>
//   render(ui, { wrapper: AllTheProviders, ...options })

// // re-export everything
// export * from "@testing-library/react"

// // override render method
// export { customRender as render }
