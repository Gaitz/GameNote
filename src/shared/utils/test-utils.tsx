import "@testing-library/jest-dom"
import { render } from "@testing-library/react"
import { Provider as ReduxProvider } from "react-redux"
import { ChakraProvider } from "@chakra-ui/react"
import store from "game-note/shared/store"
import theme from "game-note/shared/styles/theme"
import React, { FC } from "react"
import type { ReactElement } from "react"
import type { RenderOptions } from "@testing-library/react"

/* eslint-disable react/prop-types */
const AllTheProviders: FC = ({ children }) => {
  return (
    <ReduxProvider store={store}>
      <ChakraProvider theme={theme}>{children}</ChakraProvider>
    </ReduxProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
) => render(ui, { wrapper: AllTheProviders, ...options })

export * from "@testing-library/react"
export { default as userEvent } from "@testing-library/user-event"

export { customRender as render }
