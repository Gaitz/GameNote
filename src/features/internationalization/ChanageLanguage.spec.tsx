import ChangeLanguage from "./ChangeLanguage"
import { render, userEvent } from "game-note/shared/utils/test-utils"
import nookies from "nookies"

const mockPush = jest.fn()
const defaultLocale = "zh-TW"
const localeEN = "en"
const NEXT_LOCALE_COOKIE = "NEXT_LOCALE"

jest.mock("next/router", () => ({
  useRouter: () => ({
    locale: defaultLocale,
    push: mockPush
  })
}))

describe("Internationalization", () => {
  it("Snapshot", () => {
    const { asFragment } = render(<ChangeLanguage />)
    expect(asFragment()).toMatchSnapshot()
  })

  it("Change Language", () => {
    const { getByTestId } = render(<ChangeLanguage />)
    const languageSelector = getByTestId("languageSelector")

    userEvent.selectOptions(languageSelector, "")
    expect(mockPush).not.toBeCalled()

    userEvent.selectOptions(languageSelector, defaultLocale)
    expect(mockPush).not.toBeCalled()

    userEvent.selectOptions(languageSelector, localeEN)
    expect(mockPush).toBeCalled()
    expect(nookies.get(null)[NEXT_LOCALE_COOKIE]).toBe(localeEN)
  })
})
