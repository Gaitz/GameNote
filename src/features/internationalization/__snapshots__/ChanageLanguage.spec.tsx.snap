// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Internationalization Snapshot 1`] = `
<DocumentFragment>
  <div
    class="chakra-select__wrapper css-dsdwpa"
  >
    <select
      class="chakra-select css-173b75c"
      data-testid="languageSelector"
    >
      <option
        value=""
      >
        languageSwitch
      </option>
      <option
        value="en"
      >
        English
      </option>
      <option
        value="zh-TW"
      >
        繁體中文
      </option>
    </select>
    <div
      class="chakra-select__icon-wrapper css-zj2dqk"
    >
      <svg
        aria-hidden="true"
        class="chakra-select__icon"
        focusable="false"
        role="presentation"
        style="width: 1em; height: 1em; color: currentColor;"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
